### ISSUE TITLE
Peer review #2 - Looking for another independant peer review - will be compensated for your time (1 hour)

### ISSUE THREAD

**Issue by CurrencyTycoon** [CurrencyTycoon/hodl-dao#2]

**Comment by CurrencyTycoon** [comment]
All right, so peer review #1 has completed and the contract has been patched and re-deployed to Mainnet (source verified) https://etherscan.io/address/0x2926Ba12e333dDB29257F52DdE90FcE89e2E1b7D#code

For this review, please confirm that all issues in #1 have been addressed, and continue using a similar report format. Thanks.

**Comment by adamdossa** [comment]
Will do - I'll have some time this evening to start reviewing in detail.

With regards to rate - are you comfortable compensating in Ether? The job on EthLance was fixed to USD I believe.

Thanks,
Adam

________________________________

**Comment by adamdossa** [comment]
# Scope

Contract to be reviewed is hodl.sol @ 642fc92. It is deployed on mainnet at:
https://etherscan.io/address/0x2926Ba12e333dDB29257F52DdE90FcE89e2E1b7D#code

# Review

We cover the previously noted issues at:
https://github.com/CurrencyTycoon/hodl-dao/issues/1
and note any other issues.

1. `approve`

    As per previous review, this function now emits an Approve event. It also returns false if there sender does not have sufficient funds and true otherwise. It also only allows approvals for a given sender / spender pair if the existing approval for the pair is 0 (with the exception of setting the approval to 0).

    Note that there is no logic preventing Alice from approving multiple addresses, each with an approval amount equal to Alice's balance. The function prevents any single approval being greater than her balance, but does not prevent the sum of all of her approvals being greater than her balance. Some possible approaches here could be:
    - allow this to happen. I would also then remove the check:
`if (balanceOf[msg.sender] < _value) return false;`
    then semantically an approval is simply saying I approve this person to spend up to `_value` on my behalf, providing I actually have a balance. This model allows for me to approve in advance of funding my account, for example I may approve a total lifetime amount for my child to spend, but only fund my account monthly. It also allows me to have multiple equal approvees.
    - only allow one person to be approved per account, and model this with additional mappings tracking whether an existing approvee exists for an address etc..

2. `approveAndCall`

    This function has the issues previously noted for `approve`, namely:
    - ability to approve more than sender has
    - susceptible to the race condition detailed at:
https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
    - ability to approve multiple addresses with a sum of approved amounts greater than senders balance

    Probably the best approach for this function would be for it to call approve internally first, then if it gets back a true return value, call the external `receiveApproval` function, otherwise return false. This would centralise approve logic in a single function.

3. `transferFrom`

    As per previous review, modifier has been removed, and logic put into function which negates previous loophole.

    Note that the logic has also changed in that you are now not allowing a transfer if either the from or to address is pending withdrawal. This is different to the previous intended logic (whose intention was to only check that the from address was not pending withdrawal). This looks like better (simpler and cleaner) logic to me and avoids any possible issues with balances changing on accounts which have initiated withdrawals (you rely on this assumption when completing a withdrawal).

    One comment on this function (this is stylistic, so feel free to ignore ;-)) - the first 5 lines all check possible error conditions and throw as appropriate. For a non-payable function, at a point in the function where no state changes have been made, I feel it is better to return a false success response rather than throw. This is safe for these 5 lines since none of them have an effect (so there are no side-effects to roll back). This will save gas for the caller, is clearer, and easier to recover from.
    
4. `withdrawalInitiate`

    No comment.

5. `withdrawalComplete`

    As per previous review, comment has been updated to reflect the logic correctly.

    Stylistic:
    `withdrawalRequests[msg.sender].amount` could be `r.amount`
    `calculateReward(r.amount)` could be `calculateReward(amount)`

6. `calculateReward`

    As per previous review, calculation is now done correctly (multiplication first).

7. `calculateFee`

    As per previous review, calculation has been updated to `v / 100`

8. `quickWithdraw`

    As per previous review, code has been shortened to `feePot += msg.value - overAmount;`

    Note (not sure if it is intentional) is that you do not allow quick withdrawal of funds which are awaiting withdrawal. Not sure if this is intentional - personally I think the logic works better if I always have the option of a quick withdrawal with fee, regardless of what state my funds are currently in. However, note that if you do make this change you would also need to modify `withdrawalComplete` to check that the sender still has a balance greater than the amount being withdrawn before completing the withdrawal.

    Another issue is that if an insufficient fee is sent, you don't return the insufficient fee (by throwing) and instead just return false. It may be better to throw in this instance so as to return the insufficient fee (at the cost of using up all gas). On a related point, my feeling is that in general it may be better to simply throw if `(msg.value != calculateFee(amount))`, rather than refunding any excess and executing the function. Reasoning for this is that calculateFee is a constant function so is easy for any interface to calculate in order to correctly pre-populate the transaction with, and it keeps the logic a little cleaner.

9. `doWithdrawal`

    As per previous review additional check of `(amount + extra > this.balance)` has been added.

**Comment by CurrencyTycoon** [comment]
Thanks!

Really appreciate the level of detail, those are some good points.

Where/how do I send the compensation? (ETH of course)

**Comment by adamdossa** [comment]
I'll put an invoice in on EthLance if that's OK? They don't charge a
commission I think.
What amount should I invoice for?
If it's OK I'd love a rating / review on EthLance if at all possible ;-)

Good luck with the project - I think it's an interesting idea! One thought
I had is that maybe the percentage of the fee pot you get on withdrawal
should also be linked to the total time you've held your money in the
contract. You could record the time that a deposit is made, and then pay a
rate of interest (rate applying to the fee pot, not the deposit) dependent
on that - this would incentivise people to leave money there as long as
possible, similar to a fiat account.

Thanks,
Adam

**Comment by CurrencyTycoon** [comment]
Yes. My budget was 2 ETH to spend on this review. Welcome to send an invoice.

Thanks! At this point I do not want to change any of the logic as it would need re-review, but certainly a good idea. It's kind of tricky for the 'hodler', because some people would probably withdraw on purpose to get the pot - but they would need to withdraw before others - so it's like a game! On the other hand, they will have tokens, perhaps another contract can be made that does more functionality with the tokens :-)

**Comment by adamdossa** [comment]
That makes sense.

Thanks - I'll put the invoice in.

FWIW - I'd be happy to clean up the code a bit based on discussion, put
this into the truffle framework and build a basic web UI if you're
interested in getting this done.

Thanks,
Adam

**Comment by CurrencyTycoon** [comment]
Thanks for the offer, however the web UI has been done: 

See https://www.hdao.org/

New has been deployed to:  `0xd7ee73ee5a1456c1c644692685608b4b0338063d`
Source verified here: https://etherscan.io/address/0xd7ee73ee5a1456c1c644692685608b4b0338063d#code

Here is what was changed:

0. Changed to a constant instead of passing as an arguments. Supply is always intended to be 0 at the beginning. 

`decimals` is not used in this contract, but can be used by external programs, eg. wallets, to hint the wallet how to format it to the end-user.

Note that the constructor arguments are also verified on Etherscan.

1. > does not prevent the sum of all of her approvals being greater than her balance
Good point. Changed to go with the `allow this to happen` approach, commented out the check and left a comment.

2. `approveAndCall` - Changed to call the existing approve. 
Good catch. This also removes some duplicated code, always feels good to do that!

3. `transferFrom` - added comment (no code change)

Yes, it seems like exceptions are a little abused in Solidity, when looking at other examples too. Say in Java, exceptions are used for exceptional events, but in Solidity they are used for failing regular validation & rollbacks. On the other hand, if you get into the habit of `throw,` then you'll less likely get into a mistake of running in to a situation where you forgot to throw - which could be fatal. I think like a mistake like this could be made in conjunction with modifiers.

...

8. `quickWithdraw`
Yes, do not allow quick withdraw while waiting for withdrawal is intended, to keep the logic simple.
Good point with the insufficient fee. Yeah, fee can be calculated by calling the contract from UI of the DApp.
Changed to ` if (msg.value != feeRequired)` and then `throw`. Also renamed the `InsufficientFee` event to `InsufficientFee` incorrect fee.

...


I am now looking for a 3rd and final peer review, to be completed by a 3rd solidity developer. Know anyone?
